util/fee-estimator/src/estimator/confirmation_fraction.rs

```diff
@@ -0,0 +1,600 @@
+            self.tx_confirm_stat
+                .add_confirmed_tx(blocks_to_confirm, tx.fee_rate);
+            true
+        } else {
+            // tx is not tracked
+            false
+        }
+    }
+
+    /// process new block
+    /// record confirm blocks for txs which we tracked before.
+    fn process_block(&mut self, height: u64, txs: impl Iterator<Item = Byte32>) {
+        // For simpfy, we assume chain reorg will not effect tx fee.
+        if height <= self.best_height {
+            return;
+        }
+        self.best_height = height;
+        // update tx confirm stat
+        self.tx_confirm_stat.move_track_window(height);
+        self.tx_confirm_stat.decay();
+        let processed_txs = txs.filter(|tx| self.process_block_tx(height, tx)).count();
+        if self.start_height == 0 && processed_txs > 0 {
+            // start record
+            self.start_height = self.best_height;
+            ckb_logger::debug!("start recording at {}", self.start_height);
+        }
+    }
+
+    /// track a tx that entered txpool
+    fn track_tx(&mut self, tx_hash: Byte32, fee_rate: FeeRate, height: u64) {
+        if self.tracked_txs.contains_key(&tx_hash) {
+            // already in track
+            return;
+        }
+        if height != self.best_height {
+            // ignore wrong height txs
+            return;
+        }
+        if let Some(bucket_index) = self.tx_confirm_stat.add_unconfirmed_tx(height, fee_rate) {
+            self.tracked_txs.insert(
+                tx_hash,
+                TxRecord {
+                    height,
+                    bucket_index,
+                    fee_rate,
+                },
+            );
+        }
+    }
+
+    fn drop_tx_inner(&mut self, tx_hash: &Byte32, count_failure: bool) -> Option<TxRecord> {
+        self.tracked_txs.remove(tx_hash).map(|tx_record| {
+            self.tx_confirm_stat.remove_unconfirmed_tx(
+                tx_record.height,
+                self.best_height,
+                tx_record.bucket_index,
+                count_failure,
+            );
+            tx_record
+        })
+    }
+
+    /// tx removed from txpool
+    fn drop_tx(&mut self, tx_hash: &Byte32) -> bool {
+        self.drop_tx_inner(tx_hash, true).is_some()
+    }
+
+    /// estimate a fee rate for confirm target
+    fn estimate(&self, expect_confirm_blocks: BlockNumber) -> Option<FeeRate> {
+        let min_estimate_samples = constants::MIN_TARGET as usize * 2;
+        self.tx_confirm_stat.estimate_median(
+            expect_confirm_blocks as usize,
+            min_estimate_samples,
+            DEFAULT_MIN_CONFIRM_RATE,
+        )
+    }
+}
+
+impl Algorithm {
+    pub fn update_ibd_state(&mut self, in_ibd: bool) {
+        if self.is_ready {
+            if in_ibd {
+                self.clear();
+                self.is_ready = false;
+            }
+        } else if !in_ibd {
+            self.clear();
+            self.is_ready = true;
+        }
+    }
```

Comment:
`update_ibd_state` is executed every time ckb-chain finishes processing a block, is this the expected behavior?

Or should `update_ibd_state` only be executed when the `ibd_state` of ckb changes from `true` to `false`?

Reply:
I believe that when CKB enters IBD mode, `self.ibd_finished` is set to `true` only once. If CKB loses internet connection after 24 hours, `self.ibd_finished` will always remain =true=.

Cc. @driftluo

```rust
    /// Return whether chain is in initial block download
    pub fn is_initial_block_download(&self) -> bool {
        // Once this function has returned false, it must remain false.
        if self.ibd_finished.load(Ordering::Acquire) {
            false
        } else if unix_time_as_millis().saturating_sub(self.snapshot().tip_header().timestamp())
            > MAX_TIP_AGE
        {
            true
        } else {
            self.ibd_finished.store(true, Ordering::Release);
            false
        }
    }

```